src/adapter/src/coord/sequencer/inner.rs

```diff
@@ -1054,6 +1049,28 @@ impl Coordinator {
             owner_id: *session.current_role_id(),
         });
 
+        // Timestamp selection
+        let as_of = {
+            // Normally, `since` should be the least_valid_read.
+            let id_bundle = dataflow_import_id_bundle(global_lir_plan.df_desc(), cluster_id);
+            let mut as_of = self.least_valid_read(&id_bundle);
+            // But for MVs with non-trivial REFRESH schedules, it's important to set the
+            // `since` to the first refresh. This is because we'd like queries on the MV to
+            // block until the first refresh (rather than to show an empty MV).
+            if let Some(refresh_schedule) = refresh_schedule {
+                if let Some(ts) = as_of.as_option() {
+                    let mut rounded_up_as_of = Antichain::new();
+                    if let Some(rounded_up_ts) = refresh_schedule.round_up_timestamp(*ts) {
+                        rounded_up_as_of.insert(rounded_up_ts);
+                    } else {
+                        return Err(AdapterError::MaterializedViewWouldNeverRefresh(name.item));
+                    }
+                    as_of = rounded_up_as_of;
+                }
+            }
+            as_of
+        };
```

Comment:
Oopsie, good catch! I'll think this through carefully.

Btw. I'm also thinking that I should add a test that actually does a restart of envd. (My td test only restarts the replica.) Although even such a test wouldn't have caught this bug, because the bug would happen only if a crash happens at a very specific point in time.

Reply:
You might be able to trigger this today by installing the MV on a cluster with no replicas. I think the storage controller initializes the shard with an `upper = [0]` and it is only advanced from that when compute installs the `persist_sink` and advances the frontier... which it can't do without a replica.

Reply:
I think I fixed this in 0a96a639314db78b5d3b08c6a69ee44fab945c7f, but I'll think this through tomorrow again.

Reply:
Discussed with @teskje on Slack.